### PR TITLE
Support `dict()` and `{}`

### PR DESCRIPTION
**Pull request by sklam**
Implement support for `dict()` and `{}` but not dict-comprehension.

The compiler translates `dict()` to `Dict.empty(kt, vt)` where the `kt` and `vt` are types inferred based on usage.  Thus, `dict()` is still returning a `numba.typed.Dict` in jit-code.

TODO:

- [x] write docs.

### CODE REVIEW

**Comment by sklam** [comment]
It's failing on py2.7 due to the unhandled STORE_MAP

File: numba/typing/dictdecl.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+registry = Registry()
+infer = registry.register
+infer_global = registry.register_global
+infer_getattr = registry.register_attr
+
+
+_message_dict_support = """
+Unsupported usage of `dict()` with arguments. \
```
Comment (stuartarchibald) [inline numba/typing/dictdecl.py]:
```suggestion
Unsupported use of `dict()` with positional or keyword argument(s). \
```

File: numba/typing/dictdecl.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+The only implemented usage is `dict()`.
```
Comment (stuartarchibald) [inline numba/typing/dictdecl.py]:
```suggestion
The only supported use is `dict()`.
```

File: numba/typing/dictdecl.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+"""
+This implement typing template for `dict()`.
```
Comment (stuartarchibald) [inline numba/typing/dictdecl.py]:
```suggestion
This implements the typing template for `dict()`.
```

File: numba/tests/test_dicts.py [hunk 4]
```diff
@@ -22,5 +24,83 @@ def test_build_map_from_local_vars(self, flags=force_pyobj_flags):
...
+
+    def test_use_curlybraces_with_manyvar(self):
+        # Test using variable in {}
+        @njit
+        def foo(x, y):
+            return {x: 1, y: x + y}
+
+        x, y = 10, 20
+        self.assertEqual(foo(x, y), foo.py_func(x, y))
+
```
Comment (stuartarchibald) [inline numba/tests/test_dicts.py]: How about testing a mixed use case, e.g.:
```python
def foo():
  k = dict()
  k[1] = {1: 3}
  k[2] = {4: 2}
  return k
```
?

File: numba/typeinfer.py [hunk 5]
```diff
@@ -255,6 +256,31 @@ class BuildSetConstraint(_BuildContainerConstraint):
...
+            typevars = typeinfer.typevars
+            tsets = [(typevars[k.name].getone(), typevars[v.name].getone())
+                     for k, v in self.items]
+            if not tsets:
+                typeinfer.add_type(self.target,
+                                   types.DictType(types.undefined,
+                                                  types.undefined),
+                                   loc=self.loc)
+            else:
+                head = tsets[0]
```
Comment (stuartarchibald) [inline numba/typeinfer.py]: is the assignment needed here? `key_type, value_type = tsets[0]`?

Comment (sklam) [reply numba/typeinfer.py]: good catch

File: numba/tests/test_dicts.py [hunk 6]
```diff
@@ -22,5 +24,123 @@ def test_build_map_from_local_vars(self, flags=force_pyobj_flags):
...
+        with self.assertRaises(TypingError) as raises:
+            foo()
+        self.assertIn(
+            "Dict.value_type cannot be of type none",
+            str(raises.exception),
+        )
+
+
+    def test_dict_use_with_optional_value(self):
+        # Test that Optinoal cannot be used as value for Dict
```
Comment (stuartarchibald) [inline numba/tests/test_dicts.py]:
```suggestion
        # Test that Optional cannot be used as value for Dict
```

File: numba/tests/test_dicts.py [hunk 7]
```diff
@@ -22,5 +24,123 @@ def test_build_map_from_local_vars(self, flags=force_pyobj_flags):
...
+            return k
+
+        with self.assertRaises(TypingError) as raises:
+            foo(True)
+        self.assertIn(
+            "Dict.value_type cannot be of type ?float64",
+            str(raises.exception),
+        )
+
+
```
Comment (stuartarchibald) [inline numba/tests/test_dicts.py]: This also need handling:
```python
    def test_dict_use_with_optional_key(self):
        # Test that Optional cannot be used as a key for Dict
        @njit
        def foo(choice):
            k = {2.5 if choice else None: 1}
            return k

        with self.assertRaises(TypingError) as raises:
            foo(True)
```

File: examples/dict_usage.py [hunk 8]
```diff
@@ -72,3 +72,28 @@ def foo():
...
+def ex_inferred_dict_njit():
+    # magictoken.ex_inferred_dict_njit.begin
+    from numba import njit
+    import numpy as np
+
+    @njit
+    def foo():
+        d = dict()
+        k = {1: np.arange(1), 2: np.arange(2)}
+        # The following tells the compiler what the key type and the value type are
```
Comment (stuartarchibald) [inline examples/dict_usage.py]: long line

File: numba/typing/dictdecl.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+Unsupported use of `dict()` with positional or keyword argument(s). \
+The only supported use is `dict()`.
+""".strip()
+
+
+@infer_global(dict)
+class DictBuiltin(AbstractTemplate):
+    def generic(self, args, kws):
+        if args or kws:
+            raise errors.TypingError(_message_dict_support)
```
Comment (stuartarchibald) [inline numba/typing/dictdecl.py]: wondering if this should be an `UnsupportedError`?

Comment (sklam) [reply numba/typing/dictdecl.py]: We can do that but we have never done that before.  There are no raising of `UnsupportedError` from typing logic.  Also given the "extra error info" (error_rewrite) says `TypingError` is:

> This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

I'd prefer to leave it as it.

File: numba/tests/test_dicts.py (hunk 7 above)
Comment (sklam) [reply numba/tests/test_dicts.py]: add it and another one for none-key

File: numba/typing/dictdecl.py (hunk 9 above)
Comment (stuartarchibald) [reply numba/typing/dictdecl.py]: ok, this seems consistent

**Pull request merged**